zenpy/lib/request.py

```diff
@@ -231,8 +241,30 @@ def post(self, fp, token=None, target_name=None, content_type=None):
             # Other serializable types accepted by requests (like dict)
             raise ZenpyException("upload requires a target file name")
 
-        url = self.api._build_url(self.api.endpoint.upload(filename=target_name, token=token))
-        return self.api._post(url, data=fp, payload={}, content_type=content_type)
+        if isinstance(api_object, Macro) or api_object == "unassociated_macro":
+            # Macro attachment API requires attachment as well as filename
+            data = {"filename": target_name}
+            files = {"attachment": fp}
+            
+            if isinstance(api_object, Macro):
+                url = self.api._build_url(self.api.endpoint.attachments_upload(id=api_object.id))
+
+            elif api_object == "unassociated_macro":
+                url = self.api._build_url(self.api.endpoint.attachments_upload_unassociated(id=None))
+                
+            response = self.api._post(url, data=data, files=files, content_type=content_type)
+            
+        else:
+            url = self.api._build_url(self.api.endpoint.upload(filename=target_name, token=token))
+            response = self.api._post(url, data=fp, payload={}, content_type=content_type)
+
+        if not hasattr(fp, "close"):
+            raise ZenpyException("File pointer '{}' doesn't have a 'close' function.".format(fp))
```

Comment:
```
if hasattr(fp, "close"):
   fp.close()
```

Is probably better

Reply:
I agree, probably easier to read. Changed to if/else and if not available an exception is raised to ensure that file pointers always have a close method.